### ISSUE TITLE
Low mAP for custom dataset !

### ISSUE THREAD

**Issue by PumayHui** [AlexeyAB/darknet#4216]
![image](https://user-images.githubusercontent.com/28421690/68097755-c6f58d00-fef3-11e9-8f35-6e2aa90f5c7e.png)

Hi all,

I am trying to use yolov3-tiny_3l.cfg for my custom dataset with 1 classes.
I changed my .cfg for classes, filters and also obj.data files. I generated anchors for my custom dataset and put it into .cfg file.
The training and validation datasets are separate. There are no intersections between them
Cfg file as flow:

[net]
# Testing
# batch=1
# subdivisions=1
# Training
batch=64
subdivisions=16
width=1024
height=1024
channels=3
momentum=0.9
decay=0.0005
angle=0
saturation = 1.5
exposure = 1.5
hue=.1

learning_rate=0.0005
burn_in=2000
max_batches = 100000
policy=steps
steps=12074, 30200, 60400
scales=.1,.1,.1

[convolutional]
batch_normalize=1
filters=16
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=32
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=64
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=128
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=1

[convolutional]
batch_normalize=1
filters=1024
size=3
stride=1
pad=1
activation=leaky

###########

[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[convolutional]
size=1
stride=1
pad=1
filters=6
activation=linear

[yolo]
# 6,  6,   8,  7,  11,  6,   7, 11,  11,  8,  16,  7,  12, 11,  18, 18,  35, 35
mask = 8
anchors = 8,  8,  10,  9,   8, 14,  14,  8,  14, 10,  10, 14,  14, 14,  22, 23,  43, 43
classes=1
num=9
jitter=.3
ignore_thresh = .7
truth_thresh = 1
random=1

[route]
layers = -4

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[upsample]
stride=2

[route]
layers = -1, 8

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[convolutional]
size=1
stride=1
pad=1
filters=6
activation=linear

[yolo]
mask = 8
anchors = 8,  8,  10,  9,   8, 14,  14,  8,  14, 10,  10, 14,  14, 14,  22, 23,  43, 43
classes=1
num=9
jitter=.3
ignore_thresh = .7
truth_thresh = 1
random=1

[route]
layers = -3

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[upsample]
stride=2

[route]
layers = -1, 6

[convolutional]
batch_normalize=1
filters=128
size=3
stride=1
pad=1
activation=leaky

[convolutional]
size=1
stride=1
pad=1
filters=48
activation=linear

[yolo]
mask = 0,1,2,3,4,5,6,7
anchors = 8,  8,  10,  9,   8, 14,  14,  8,  14, 10,  10, 14,  14, 14,  22, 23,  43, 43
classes=1
num=9
jitter=.3
ignore_thresh = .7
truth_thresh = 1
random=1

I can see the loss going down, but the mAP is low very much.(see graph with mAP)

How can I solve this problem?Any suggestions..?
Thanks.

**Comment by PumayHui** [comment]
And I checked using Yolo mark, it shows correct BB on images.
When I run this command:
`./darknet` detector calc_anchors data/obj.data -num_of_clusters 9 -width 1024 -height 1024 -show'
The cloud.png as follow:
![image](https://user-images.githubusercontent.com/28421690/68097938-d32e1a00-fef4-11e9-8dc1-27014a9f46bd.png)

**Comment by PumayHui** [comment]
And because I have 38635 training images and set batch=64, then epoch = 38635/64 = 604 iterations
So
20 epochs = 12074 iterations
50 epochs = 30200 iterations
100 epochs = 60400 iterations

So I have Set
`
 steps=12074, 30200, 60400
 scales=.1,.1,.1
`
.

**Comment by AlexeyAB** [comment]
Try to train with default anchors

**Comment by lsd1994** [comment]
It seems the object you want to detect is very small, right?

**Comment by PumayHui** [comment]
Yep @lsd1994

**Comment by lsd1994** [comment]
@PumayHui So your dataset is very difficult, bad performance is normal.

**Comment by PumayHui** [comment]
@lsd1994 But this issue gets better results with small objects ......
[https://github.com/AlexeyAB/darknet/issues/2960](url)

**Comment by AlexeyAB** [comment]
Try to train with default anchors
Also what is the average size of your training images?

**Comment by PumayHui** [comment]
@AlexeyAB The average size of my training images is (1024, 1024).

**Comment by AlexeyAB** [comment]
Use this cfg-file: https://github.com/AlexeyAB/darknet/blob/master/cfg/yolov3-tiny_3l.cfg
with default anchors and width=1024 height=1024

**Comment by PumayHui** [comment]
When I use default anchors and width=1024 height=1024 in yolov3-tiny_3l.cfg, mAP png as follow:
![image](https://user-images.githubusercontent.com/28421690/68908586-c1ffcb80-0786-11ea-8c87-a940af405d79.png)
![image](https://user-images.githubusercontent.com/28421690/68908610-ce842400-0786-11ea-92b0-a7bcbe52ac33.png)
Interrupted due to network reasons.
Could you give me some advices? Thanks so much!

**Comment by AlexeyAB** [comment]
Try to follow: https://github.com/AlexeyAB/darknet#how-to-train-to-detect-your-custom-objects
change line steps to 80% and 90% of max_batches, f.e. steps=4800,5400

And follow these rules: https://github.com/AlexeyAB/darknet#how-to-improve-object-detection
